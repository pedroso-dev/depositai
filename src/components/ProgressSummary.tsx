import { formatCurrency } from "../utils/formatCurrency";

type ProgressSummaryProps = {
  completedCount: number;
  totalCount: number;
  totalSaved: number;
  totalGoal: number;
  progressPercentage: number;
  onReset: () => void;
};

export default function ProgressSummary({
  completedCount,
  totalCount,
  totalSaved,
  totalGoal,
  progressPercentage,
  onReset,
}: ProgressSummaryProps) {
  if (progressPercentage > 100) {
    progressPercentage = 100;
  }

  return (
    <div className="mb-8 w-full max-w-4xl rounded-lg bg-gray-800 p-6 shadow-lg">
      <div className="mb-4 flex items-center justify-between text-white">
        <div className="flex flex-col">
          <span className="text-lg font-bold" data-testid="total-saved">
            {formatCurrency(totalSaved)}
          </span>
          <span className="text-sm text-gray-400" data-testid="total-goal">
            de {formatCurrency(totalGoal)}
          </span>
        </div>
        <div className="text-right" data-testid="deposit-counts">
          <span className="text-lg font-bold">{completedCount}</span>
          <span className="text-sm text-gray-400">/{totalCount} depósitos</span>
        </div>
      </div>
      <div
        className="h-4 w-full rounded-full bg-gray-600"
        role="progressBar"
        aria-label="Progresso do desafio"
        aria-valuenow={totalSaved}
        aria-valuemin={0}
        aria-valuemax={totalGoal}
      >
        <div
          className="h-4 rounded-full bg-green-500 transition-all duration-300"
          style={{ width: `${progressPercentage}%` }}
        ></div>
      </div>
      <div className="mt-6 text-center">
        <button
          className="rounded-md bg-red-800 px-4 py-2 text-sm font-semibold text-white transition-colors hover:bg-red-700 focus:outline-none focus:ring-2 focus:ring-red-500"
          onClick={onReset}
        >
          Apagar desafio e recomeçar
        </button>
      </div>
    </div>
  );
}
