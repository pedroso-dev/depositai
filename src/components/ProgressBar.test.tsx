import { render, screen } from "@testing-library/react";
import ProgressSummary from "./ProgressSummary";

describe("<ProgressSumary />", () => {
  it("should render the summary with partial progress", () => {
    const props = {
      completedCount: 25,
      totalCount: 100,
      totalSaved: 1250,
      totalGoal: 5050, // Soma de 1 a 100
    };
    const progressPercentage = (props.totalSaved / props.totalGoal) * 100;
    render(
      <ProgressSummary progressPercentage={progressPercentage} {...props} />
    );
    expect(screen.getByText(/R\$\s*1\.250,00/i)).toBeInTheDocument();
    expect(screen.getByText(/de R\$\s*5\.050,00/i)).toBeInTheDocument();
    expect(screen.getByTestId("deposit-counts")).toHaveTextContent(
      "25/100 depósitos"
    );

    const progressBar = screen.getByRole("progressBar");
    expect(progressBar).toBeInTheDocument();
    expect(progressBar).toHaveAttribute("aria-valuenow", "1250");
    expect(progressBar).toHaveAttribute("aria-valuemax", "5050");

    const progressBarFill = progressBar.firstChild as HTMLElement;
    expect(progressBarFill).toHaveStyle(`width: ${progressPercentage}%`);
  });

  it("should render correctly for a challenge at 0% progress", () => {
    const props = {
      completedCount: 0,
      totalCount: 200,
      totalSaved: 0,
      totalGoal: 20100,
    };
    const progressPercentage = (props.totalSaved / props.totalGoal) * 100;
    render(
      <ProgressSummary progressPercentage={progressPercentage} {...props} />
    );

    expect(screen.getByText(/R\$\s*0,00/i)).toBeInTheDocument();
    expect(screen.getByTestId("deposit-counts")).toHaveTextContent(
      "0/200 depósitos"
    );

    const progressBarFill = screen.getByRole("progressBar")
      .firstChild as HTMLElement;
    expect(progressBarFill).toHaveStyle("width: 0%");
  });

  it("should render correctly for a completed challenge at 100%", () => {
    const props = {
      completedCount: 50,
      totalCount: 50,
      totalSaved: 1275,
      totalGoal: 1275,
    };
    const progressPercentage = (props.totalSaved / props.totalGoal) * 100;
    render(
      <ProgressSummary progressPercentage={progressPercentage} {...props} />
    );

    const totalSavedElement = screen.getByTestId("total-saved");
    const totalGoalElement = screen.getByTestId("total-goal");
    expect(totalSavedElement).toHaveTextContent("R$ 1.275,00");
    expect(totalGoalElement).toHaveTextContent("de R$ 1.275,00");
    expect(screen.getByTestId("deposit-counts")).toHaveTextContent(
      "50/50 depósitos"
    );

    const progressBarFill = screen.getByRole("progressBar")
      .firstChild as HTMLElement;
    expect(progressBarFill).toHaveStyle("width: 100%");
  });

  it("should handle a totalGoal of 0 to avoid division by zero", () => {
    const props = {
      completedCount: 0,
      totalCount: 0,
      totalSaved: 0,
      totalGoal: 0,
    };
    const progressPercentage = (props.totalSaved / props.totalGoal) * 100;
    render(
      <ProgressSummary progressPercentage={progressPercentage} {...props} />
    );
    const progressBarFill = screen.getByRole("progressBar");
    expect(progressBarFill).toHaveAttribute("aria-valuenow", "0");
    expect(progressBarFill).toHaveAttribute("aria-valuemax", "0");
  });

  it("should handle floating-point numbers correctly", () => {
    const props = {
      completedCount: 1,
      totalCount: 2,
      totalSaved: 150.75,
      totalGoal: 301.5,
    };
    const progressPercentage = (props.totalSaved / props.totalGoal) * 100;
    render(
      <ProgressSummary progressPercentage={progressPercentage} {...props} />
    );
    expect(screen.getByTestId("total-saved")).toHaveTextContent("R$ 150,75");
    expect(screen.getByTestId("total-goal")).toHaveTextContent("de R$ 301,50");
    const progressBarFill = screen.getByRole("progressBar")
      .firstChild as HTMLElement;
    expect(progressBarFill).toHaveStyle("width: 50%");
  });

  it("should cap the progress bar at 100% even if saved amount exceeds goal", () => {
    const props = {
      completedCount: 10,
      totalCount: 5,
      totalSaved: 2000,
      totalGoal: 1000,
    };
    const progressPercentage = (props.totalSaved / props.totalGoal) * 100;
    render(
      <ProgressSummary progressPercentage={progressPercentage} {...props} />
    );

    expect(screen.getByTestId("total-saved")).toHaveTextContent("R$ 2.000,00");
    expect(screen.getByTestId("total-goal")).toHaveTextContent(
      "de R$ 1.000,00"
    );

    const progressBarFill = screen.getByRole("progressBar")
      .firstChild as HTMLElement;
    expect(progressBarFill).toHaveStyle("width: 100%");
  });

  it("should update correctly when props change", () => {
    const initialProps = {
      completedCount: 50,
      totalCount: 100,
      totalSaved: 500,
      totalGoal: 1000,
    };
    const progressPercentageInitial =
      (initialProps.totalSaved / initialProps.totalGoal) * 100;
    const { rerender } = render(
      <ProgressSummary
        progressPercentage={progressPercentageInitial}
        {...initialProps}
      />
    );
    expect(screen.getByTestId("total-saved")).toHaveTextContent("R$ 500,00");
    expect(
      screen.getByRole("progressBar").firstChild as HTMLElement
    ).toHaveStyle("width: 50%");

    const updatedProps = {
      ...initialProps,
      completedCount: 75,
      totalSaved: 750,
    };
    const progressPercentageUpdated =
      (updatedProps.totalSaved / updatedProps.totalGoal) * 100;
    rerender(
      <ProgressSummary
        progressPercentage={progressPercentageUpdated}
        {...updatedProps}
      />
    );
    expect(screen.getByTestId("total-saved")).toHaveTextContent("R$ 750,00");
    expect(
      screen.getByRole("progressBar").firstChild as HTMLElement
    ).toHaveStyle("width: 75%");
  });

  it("should match the snapshot for a typical progress state", () => {
    const props = {
      completedCount: 25,
      totalCount: 100,
      totalSaved: 1250,
      totalGoal: 5050,
    };
    const progressPercentage = (props.totalSaved / props.totalGoal) * 100;
    const { container } = render(
      <ProgressSummary progressPercentage={progressPercentage} {...props} />
    );

    expect(container).toMatchSnapshot();
  });
});
