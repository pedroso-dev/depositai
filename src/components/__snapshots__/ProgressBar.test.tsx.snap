// Jest Snapshot v1, https://jestjs.io/docs/snapshot-testing

exports[`<ProgressSumary /> should match the snapshot for a typical progress state 1`] = `
<div>
  <div
    class="mb-8 w-full max-w-4xl rounded-lg bg-gray-800 p-6 shadow-lg"
  >
    <div
      class="mb-4 flex items-center justify-between text-white"
    >
      <div
        class="flex flex-col"
      >
        <span
          class="text-lg font-bold"
          data-testid="total-saved"
        >
          R$ 1.250,00
        </span>
        <span
          class="text-sm text-gray-400"
          data-testid="total-goal"
        >
          de 
          R$ 5.050,00
        </span>
      </div>
      <div
        class="text-right"
        data-testid="deposit-counts"
      >
        <span
          class="text-lg font-bold"
        >
          25
        </span>
        <span
          class="text-sm text-gray-400"
        >
          /
          100
           depósitos
        </span>
      </div>
    </div>
    <div
      aria-label="Progresso do desafio"
      aria-valuemax="5050"
      aria-valuemin="0"
      aria-valuenow="1250"
      class="h-4 w-full rounded-full bg-gray-600"
      role="progressBar"
    >
      <div
        class="h-4 rounded-full bg-green-500 transition-all duration-300"
        style="width: 24.752475247524753%;"
      />
    </div>
    <div
      class="mt-6 text-center"
    >
      <button
        class="rounded-md bg-red-800 px-4 py-2 text-sm font-semibold text-white transition-colors hover:bg-red-700 focus:outline-none focus:ring-2 focus:ring-red-500"
      >
        Apagar desafio e recomeçar
      </button>
    </div>
  </div>
</div>
`;
